rotkehlchen/chain/optimism/modules/velodrome/decoder.py

```diff
@@ -0,0 +1,83 @@
+from rotkehlchen.chain.evm.types import string_to_evm_address
+from rotkehlchen.chain.optimism.modules.velodrome.constants import CPT_VELODROME
+from rotkehlchen.chain.optimism.modules.velodrome.velodrome_cache import (
+    read_velodrome_pools_and_gauges_from_cache,
+)
+from rotkehlchen.types import ChecksumEvmAddress
+
+if TYPE_CHECKING:
+    from rotkehlchen.chain.evm.decoding.base import BaseDecoderTools
+    from rotkehlchen.chain.optimism.manager import OptimismInquirer
+    from rotkehlchen.user_messages import MessagesAggregator
+
+
+ROUTER_V2 = string_to_evm_address('0xa062ae8a9c5e11aaa026fc2670b0d65ccc8b2858')
+ADD_LIQUIDITY_EVENT = b'L \x9b_\xc8\xadPu\x8f\x13\xe2\xe1\x08\x8b\xa5jV\r\xffi\n\x1co\xef&9OL\x03\x82\x1cO'  # Mint event (mints LP tokens) # 0x4c209b5fc8ad50758f13e2e1088ba56a560dff690a1c6fef26394f4c03821c4f  # noqa: E501
+REMOVE_LIQUIDITY_EVENT = b']bJ\xa9\xc1H\x15:\xb3Dl\x1b\x15Of\x0e\xe7p\x1eT\x9f\xe9\xb6-\xabqq\xb1\xc8\x0eo\xa2'  # Burn event (burns LP tokens) # 0x5d624aa9c148153ab3446c1b154f660ee7701e549fe9b62dab7171b1c80e6fa2  # noqa: E501
+SWAP_FROM_OR_TO_ETH = b'\xb3\xe2w6\x06\xab\xfd6\xb5\xbd\x919K:T\xd19\x836\xc6P\x05\xba\xf7\xbfz\x05\xef\xef\xfa\xf7['  # It can be any eth (WETH, frETH etc.)  # 0xb3e2773606abfd36b5bd91394b3a54d1398336c65005baf7bf7a05efeffaf75b  # noqa: E501
+SWAP_TOKEN_TO_TOKEN = b'\xd7\x8a\xd9_\xa4l\x99KeQ\xd0\xda\x85\xfc\'_\xe6\x13\xce7e\x7f\xb8\xd5\xe3\xd10\x84\x01Y\xd8"'  # 0xd78ad95fa46c994b6551d0da85fc275fe613ce37657fb8d5e3d130840159d822  # noqa: E501
+# some notes:
+# swap events have at least one log with the address of the pool the swap happened with.
+# the eth amount swapped is in the internal transactions
+
+# next steps:
+# add LP to gauge (deposit LP tokens to gauge)
+# remove LP from gauge (withdraw LP tokens from gauge)
+
+
+class VelodromeDecoder(DecoderInterface, ReloadableDecoderMixin):
+
+    def __init__(
+            self,
+            optimism_inquirer: 'OptimismInquirer',
+            base_tools: 'BaseDecoderTools',
+            msg_aggregator: 'MessagesAggregator',
+    ) -> None:
+        super().__init__(
+            evm_inquirer=optimism_inquirer,
+            base_tools=base_tools,
+            msg_aggregator=msg_aggregator,
+        )
+        self.curve_pools, self.curve_gauges = read_velodrome_pools_and_gauges_from_cache()
+        self.inquirer = optimism_inquirer
+
+    def _decode_velodrome_events(self) -> None:
+        pass
+
+    def _decode_velodrome_gauge_events(self) -> None:
+        pass
+
+    def counterparties(self) -> list['CounterpartyDetails']:
+        return [CounterpartyDetails(identifier=CPT_VELODROME, label='velodrome_finance', image='velodrome.svg')]  # noqa: E501
+
+    def reload_data(self) -> Optional[Mapping[ChecksumEvmAddress, tuple[Any, ...]]]:
+        """Make sure velodrome pools are recently queried from the chain, saved in the DB
+        and loaded to the decoder's memory.
```

Comment:
if the logic is the same as in curve (i believe is almost identical) maybe you can abstract it

Reply:
It depends on what new mappings have to be added. The decoding logic is not written yet.